mp4parse/src/lib.rs

```diff
@@ -160,32 +165,205 @@ pub fn extend_from_slice<T: Clone>(vec: &mut Vec<T>, other: &[T]) -> std::result
+        /// around `std::io::Read::read_to_end()`.
+        fn try_read_to_end(&mut self, buf: &mut TryVec<u8>) -> Result<usize, ()> {
+            try_read_up_to(self, self.limit(), buf)
+        }
+    }
+
+    fn try_read_up_to<R: Read>(src: &mut R, limit: u64, buf: &mut TryVec<u8>) -> Result<usize, ()> {
+        let additional = limit.try_into().map_err(|_| ())?;
+        buf.reserve(additional)?;
+        let bytes_read = src.read_to_end(&mut buf.inner).map_err(|_| ())?;
         Ok(bytes_read)
     }
-    #[cfg(not(feature = "mp4parse_fallible"))]
-    {
-        src.read_to_end(buf).map_err(|_| ())
+
+    #[derive(Debug, Default)] // TODO: make debug print like Vec instead of deriving
+    pub struct TryVec<T> {
+        inner: Vec<T>,
+    }
+
+    impl<T> TryVec<T> {
+        pub fn new() -> Self {
+            Self { inner: Vec::new() }
+        }
+
+        pub fn with_capacity(capacity: usize) -> Result<Self, ()> {
+            Ok(Self {
+                inner: vec_with_capacity(capacity)?,
+            })
+        }
+
+        pub fn append(&mut self, other: &mut Self) -> Result<(), ()> {
+            self.reserve(other.inner.len())?;
+            self.inner.append(&mut other.inner);
+            Ok(())
+        }
+
+        // TODO remove when all Vecs are converted to TryVec
+        pub fn into_inner(self) -> Vec<T> {
+            self.inner
+        }
+
+        pub fn is_empty(&self) -> bool {
+            self.inner.is_empty()
+        }
+
+        pub fn iter_mut(&mut self) -> IterMut<T> {
+            IterMut {
+                inner: self.inner.iter_mut(),
+            }
+        }
+
+        pub fn iter(&self) -> Iter<T> {
+            Iter {
+                inner: self.inner.iter(),
+            }
+        }
+
+        pub fn push(&mut self, value: T) -> Result<(), ()> {
+            vec_push(&mut self.inner, value)
+        }
+
+        fn reserve(&mut self, additional: usize) -> Result<(), ()> {
+            #[cfg(feature = "mp4parse_fallible")]
+            {
+                FallibleVec::try_reserve(&mut self.inner, additional)
+            }
+            #[cfg(not(feature = "mp4parse_fallible"))]
+            {
+                self.inner.reserve(additional);
+                Ok(())
+            }
+        }
+    }
+
+    impl<T: Clone> TryVec<TryVec<T>> {
+        pub fn concat(&self) -> Result<TryVec<T>, ()> {
+            let size = self.iter().map(|v| v.inner.len()).sum();
+            let mut result: TryVec<T> = TryVec::with_capacity(size)?;
+            for v in self.iter() {
+                result.extend_from_slice(&v.inner)?;
+            }
+            Ok(result)
+        }
+    }
+
+    impl<T: Clone> TryVec<T> {
+        pub fn extend_from_slice(&mut self, other: &[T]) -> Result<(), ()> {
+            extend_from_slice(&mut self.inner, other)
+        }
+    }
+
+    impl<T: Clone> Clone for TryVec<T> {
+        fn clone(&self) -> Self {
+            Self {
+                inner: self.inner.clone(),
```

Comment:
This is tricky, since `Clone` has no way to signal failure, but the implied allocation in `self.inner.clone()` would ideally be fallible.

Can we get away with leaving this unimplemented and deliberately avoiding uses, or is that going to be too messy?

Reply:
Good catch! So far, it looks like we can get away with avoiding uses and I've updated the PR with a fix: see https://github.com/mozilla/mp4parse-rust/pull/211/commits/54aa1e594c9763e122bbb46be6fdabee24e493d5

I also noticed a similar bug in `extend_from_slice` which is fixed.